ambry-api/src/main/java/com.github.ambry/config/ConnectionPoolConfig.java

```diff
@@ -60,6 +60,14 @@
   @Default("2")
   public final int connectionPoolMaxConnectionsPerPortSSL;
 
+  /**
+   * The maximum consecutive connect failures allowed before we clean up the available pool
+   */
+  @Config("connectionpool.max.connect.failures.to.clean.up.available.pool")
```

Comment:
@sriramsub about the impact: the issue we are trying to fix here was detected when we increased the connection pool size from 5 to 20 (which was required with the coordinator backed router to handle peak time loads). What we saw was that if a server node went down, the frontend would mark the node down for a long time. We debugged and found out that this is due to the following: let us say at some point there were 20 established connections to a server. When the server goes down with requests in flight, the coordinator calls `destroyConnection()` on those connections. Once 3 such failures are received, the resource is marked down by the failure detector. However, the other 17 connections or so are still valid as far as the `BlockingChannelConnectionPool` is concerned (although they are actually stale). So even when the server is back up, and the failure detector conditionally marks the node as up, a subsequent request for that server tries to reuse a stale connection rather than creating a new connection. This fails the request, and the failure detector immediately marks the node as down for another 10 minutes. With an increase in the connection pool size, this situation gets exacerbated.

This is not something that can be fixed by the failure detector really, as the fault is due to using stale connections. There are two ways to fix this that I can think of:
1.  In `destroyConnection()` when the `BlockingChannel` tries to recreate a connection to a node and fails, destroy all available connections to that host port (because the chances of them being valid is slim, and in the worst case new connections will be established).
2. In the coordinator when a failed response is received and the failure detector is notified, immediately check with the failure detector about the status of the node and if the node is down, then call into `BlockingChannel` to close all connections to the node.

1 is what this patch is trying to do (I am just of the opinion that we shouldn't have a retry count). 2 would require an api addition.

Also, note that this is all temporary. The selector way of doing things does not have this problem as the router proactively polls for events on every key.